tfutils/base.py

```diff
@@ -816,49 +950,112 @@ def train(sess,
+
+        for (step, trarg, train_res) in zip(steps, trargs, train_results):
+
+            old_step = step
+            step = trarg['global_step'].eval(session=trarg['sess'])
+
+            if step <= old_step:
+                raise NoChangeError('Your optimizer should have incremented the global step,'
+                                    ' but did not: old_step=%d, new_step=%d' % (old_step, step))
+            if train_res['loss'] > trarg['thres_loss']:
+                raise HiLossError('Loss {:.2f} exceeded the threshold {:.2f}'.format(train_res['loss'],
+                                                                                     trarg['thres_loss']))
 
-    if step >= num_steps:
-        log.info('Training cancelled since step (%d) is >= num_steps (%d)' % (step, num_steps))
-        return
+            # Validation
+            vtargs = trarg['validation_targets'] if step % trarg['dbinterface'].save_valid_freq == 0 else {}
+            valid_res = run_targets_dict(trarg['sess'], vtargs)
 
-    log.info('Training beginning ...')
-    coord, threads = start_queues(sess)
+            # Save
+            trarg['dbinterface'].start_time_step = start_time_step
+            trarg['dbinterface'].save(train_res=train_res,
+                                      valid_res=valid_res,
+                                      validation_only=False)
 
-    if step == 0:
-        dbinterface.start_time_step = time.time()
-        if validate_first:
-            validation_res = run_targets_dict(sess, validation_targets,
-                                          dbinterface=dbinterface)
-    while step < num_steps:
-        old_step = step
-        dbinterface.start_time_step = time.time()
-        train_results = sess.run(train_targets)
-        step = global_step.eval(session=sess)
-        if step <= old_step:
-            raise NoChangeError('Your optimizer should have incremented the global step,'
-                                ' but did not: old_step=%d, new_step=%d' % (old_step, step))
-        if train_results['loss'] > thres_loss or np.isnan(train_results['loss']):
-            raise HiLossError('Loss {:.2f} exceeded the threshold {:.2f}'.format(train_results['loss'], thres_loss))
+        steps = [t['global_step'].eval(session=t['sess']) for t in trargs]
 
-        vtargs = validation_targets if step % dbinterface.save_valid_freq == 0 else {}
-        validation_res = run_targets_dict(sess, vtargs)
-        dbinterface.save(train_res=train_results, valid_res=validation_res, validation_only=False)
+    # Sync and close the session
+    res = []
+    for trarg in trargs:
+        stop_queues(trarg['sess'], trarg['queues'],
+                    trarg['coord'], trarg['threads'])
+        trarg['dbinterface'].sync_with_host()
+        res.append(trarg['dbinterface'].outrecs)
 
-    stop_queues(sess, queues, coord, threads)
-    dbinterface.sync_with_host()
-    return dbinterface.outrecs
+    trargs[0]['sess'].close()
```

Comment:
how come we just get away with closing trargs[0]['sess']?  are we assuming the sess is the same for all trargs elements